Allow jsonlines as input (#271)
EducationalTestingService/rsmtool: doc/pipeline.rst

Input file format
"""""""""""""""""

The input files containing feature values and scores for all responses in training and evaluation data should be either in tabular format with features and scores stored in columns and each row correponding to a single response or in ``jsonlines`` format with one line per response. See below for further description of ``.jsonlines`` format. 

RSMTool supports input files in ``.csv``, ``.tsv``, ``.sas7bdat``, ``xls``/``.xlsx``, or ``.jsonlines`` format. The format of the file is determined based on the extension. In all cases the :ref:`output files<intermediate_files_rsmtool>` will be saved in ``.csv`` format by default (see :ref:`file format <file_format>` for other intermediate file output options).

For ``Excel`` spreadsheets all data must be stored in the first sheet. 

> Each line in ``.jsonlines`` format should correspond to one response represented as a dictionary with column names as keys and values as values. For example:

Reviewer: ```suggestion
In a ``.jsonlines`` format file, each line corresponds to a response and is represented as a dictionary with column names as the keys and column values as the values. For example:
```

.. literalinclude:: ../tests/data/files/train.jsonlines
   :lines: 1,2

> RSMTool allows nested ``jsons`` in each line of ``.jsonlines`` file with no more than one level of nesting as in the following example. Note that only the lower level keys will be used when parsing the files and processing the data. Thus in the example below second level labels ``x`` and ``metadata`` will be ignored. 

Reviewer: ```suggestion
Although RSMTool does allow for a single level of nesting in the JSON objects on each line of a ``.jsonlines`` format file, *only* the top-level keys are actually used when parsing the files and processing the data. Therefore, in the example bellow, the keys ``x`` and ``metadata`` will be ignored. 
```

Input file format
"""""""""""""""""

The input files for the training and evaluation data should either be in a tabular format with responses as rows and features and score(s) in the columns  in the ``jsonlines`` format with a JSON object per responses on each line. See below for a more detailed description of the ``.jsonlines`` format. 

RSMTool supports input files in ``.csv``, ``.tsv``, ``.sas7bdat``, ``xls``/``.xlsx``, or ``.jsonlines`` format. The format of the file is determined based on the extension. In all cases the :ref:`output files<intermediate_files_rsmtool>` will be saved in ``.csv`` format by default (see :ref:`file format <file_format>` for other intermediate file output options).

For ``Excel`` spreadsheets, all data must be stored in the first sheet. 

In a ``.jsonlines`` format file, each line corresponds to a response and is represented as a dictionary with column names as the keys and column values as the values. For example:

.. literalinclude:: ../tests/data/files/train.jsonlines
   :lines: 1,2

> Although RSMTool does allow for nesting in the JSON objects on each line of a ``.jsonlines`` format file, the top-level keys will be ignored when parsing the files and processing the data. Therefore, in the example bellow, the keys ``x`` and ``metadata`` will be ignored. 

Reviewer: ```suggestion
Although RSMTool does allow for nesting in the JSON objects on each line of a ``.jsonlines`` format file, the top-level keys will be ignored when parsing the files and processing the data. Therefore, in the example below, the keys ``x`` and ``metadata`` will be ignored. 
```

.. literalinclude:: ../tests/data/files/train_nested.jsonlines
   :lines: 1,2

> If the data contains nested records with more than two levels, the column names for nested records beyond the top level will be generated using ``.`` to separate levels: ``{'foo': {'bar': {'foo2': 0, 'foo3': 0}}`` -> ``foo`` will be ignored, the columns will be named ``bar.foo2`` and ``bar.foo3``.

Reviewer: ```suggestion
If the file contains nesting of more than two levels, the column names for nested records beyond the top level will be generated using ``.`` to separate levels: For example, given the JSON object ``{'foo': {'bar': {'foo2': 0, 'foo3': 0}}``,  ``foo`` will be ignored and the columns will be named ``bar.foo2`` and ``bar.foo3``.
```
